---
title: "species-id-jan-2024"
author: "diana baetscher"
date: "2024-01-02"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# suppress R warnings for packages
defaultW <- getOption("warn")
options(warn = -1)
```

Processing rockfish species ID data from 2023-11-21.


Walkthrough of species id for unknown samples:

Using the rds file output from microhaplot:
1. read in rds files
2. apply read depth filters
3. apply allele balance filter


```{r install-packages-and-load-functions}
library(tidyverse)
library(readxl)
library(stringr)
library(lubridate)
library(rubias)
library(ggpattern)
# for mapping
library(ggplot2)
library(sf)  
library(rnaturalearth)
library(rnaturalearthdata)




source("../R/rockfish-funcs2.R")


#### Call genos from the microhaplot rds files ####

# the directory with the rds file
dir <- "../data_AFSC/rds_files/"


# cycle over them, read them and add the gtseq_run column on each.
# at the end, bind them together.
genos_long <- call_genos_from_haplotRDS(path = file.path(dir, "rockfish_multiSPP_id.rds"))

#### In the end, let us get a data frame that includes genotypes for all the individuals  ####
# and which explicitly has NAs in places where data are missing
genos_long_explicit_NAs <- genos_long %>%
  select(id) %>%
  unique() %>%
  unlist() %>%
  unname() %>%
  expand.grid(id = ., locus = unique(genos_long$locus), gene_copy = 1:2, stringsAsFactors = FALSE) %>%
  tbl_df() %>% 
  left_join(., genos_long) %>%
  arrange(id, locus, gene_copy)

  
```


Using those genotypes...
```{r}
genos_long_explicit_NAs %>%
  group_by(id) %>%
  tally()

```

288 samples.


Look at missing data:
180 gene copies total (90 loci x2)
```{r}
ind_to_toss <- genos_long_explicit_NAs %>%
  group_by(id) %>%
  filter(is.na(allele)) %>% # missing data
  tally() %>%
  arrange(desc(n)) %>% # remove samples with >20% missing data
  filter(n > 36) 

# remove those from the df
genos_ind_filtered <- genos_long_explicit_NAs %>%
  anti_join(., ind_to_toss)
  
```

21 samples to toss because of >20% missing data.

Load baseline data
```{r baseline-data}
# baseline data - curated, 997 indivs
baseline <- readRDS("../new_baseline_data/processed/sebastes_spp_id_baseline_haplotypes.rds")

# remove the 6 loci that had HWE and other issues
to_remove <- read_csv("../data/loci_to_remove.csv")

baseline90 <- baseline %>%
  anti_join(., to_remove)

# remind myself which species are in the baseline:
baseline90 %>%
  select(collection) %>%
  unique() %>%
  arrange()

tossers <- baseline90 %>%
  select(indiv, gtseq_run, id) %>%
  unique() %>%
  group_by(indiv) %>%
  tally() %>%
  filter(n >1)

baseline90_one_each <- baseline90 %>%
  anti_join(., tossers)

# baseline data - curated, 997 indivs
baseline_spp_info <- baseline90_one_each %>%
  select(sample_type, repunit, collection, indiv, gtseq_run, id, species) %>%
  unique()
baseline_spp_info$gtseq_run <- as.character(baseline_spp_info$gtseq_run)

```



```{r baseline-format}
# slim that down to just the matching field with the unknowns
for_alleidx <- baseline90_one_each %>%
  select(-indiv, -c(1:3, 12:13), -species)
  

for_alleidx$gtseq_run <- as.character(for_alleidx$gtseq_run)
```



```{r}
# merge the two dataframes
merged_df <- bind_rows(for_alleidx, genos_ind_filtered)

# first make integers of the alleles
alle_idxs <- merged_df %>% 
  dplyr::select(gtseq_run, id, locus, gene_copy, allele) %>%
  group_by(locus) %>%
  mutate(alleidx = as.integer(factor(allele, levels = unique(allele)))) %>%
  ungroup() %>%
  arrange(gtseq_run, id, locus, alleidx) # rubias can handle NA's, so no need to change them to 0's

  
# and spread the alleles
two_col <- alle_idxs %>%
  #group_by(indiv, locus) %>%
  unite(loc, locus, gene_copy, sep = ".") %>%
  #ungroup() %>%
  select(-allele) %>%
  pivot_wider(names_from = loc, values_from = alleidx) 


```



add back on info for reference and make two-column format for rubias
```{r}
# baseline
reference <- two_col %>%
  left_join(., baseline_spp_info) %>%
  filter(!is.na(species)) %>%
  select(-gtseq_run, -id, -species) %>%
  select(sample_type, repunit, collection, indiv, everything())

```

```{r}
# mixture
rubias_mix <- two_col %>%
  anti_join(., baseline_spp_info) %>%
  select(-gtseq_run) %>%
  mutate(sample_type = "mixture", collection = "spp_id", repunit = NA) %>%
  select(sample_type, repunit, collection, everything()) %>%
  rename(indiv = id)
```


## Mixture assignment with rubias


```{r run-rubias}
rubias_output <- infer_mixture(reference = reference, mixture = rubias_mix, gen_start_col = 5)

# we know there are species sequenced that aren't currently in the baseline.
# those should come out in the z-scores
```


```{r individual-posteriors}
# take the top output for each sample
top_assign <- rubias_output$indiv_posteriors %>%
  group_by(indiv) %>%
  slice_max(., order_by = PofZ)


top_assign 
```


```{r summary}
top_assign %>%
  filter(z_score < -3 | z_score > 3)

```
A bunch of funky z-scores




### add metadata

```{r read-in-sample-list}

# sample list
ss <- read_tsv("../data_AFSC/samples_list.txt", col_names = F)
  
# merge the ids with some of the new species
results_w_spp <- as.data.frame(str_split_fixed(ss$X1, "_", 3)) %>%
  select(V1, V2) %>%
  rename(sample_id = V1, indiv = V2) %>%
  mutate(indiv = str_replace(indiv, "S", "s")) %>%
  unique() %>% # because there are two entries for each samples (R1 and R2)
  left_join(., top_assign)


results_w_spp %>%
  filter(str_detect(sample_id, "ABLG")) %>%
   filter(z_score < -3 | z_score > 3)

```
That's close enough to include.

So the distribution of unknowns is:
```{r}
results_w_spp %>%
  filter(str_detect(sample_id, "ABLG")) %>%
  group_by(repunit) %>%
  tally()

```


The rest of the samples should be species to add to the baseline:
```{r}
results_w_spp %>%
  filter(!str_detect(sample_id, "ABLG")) %>%
  filter(z_score > -3 | z_score < 3) # within expected z-score distributions


```

Outputs:
```{r}
output <- results_w_spp %>%
  filter(z_score > -3 & z_score < 3.1) %>% # within expected z-score distributions
  select(sample_id, repunit, PofZ, z_score, n_non_miss_loci, n_miss_loci) %>%
  rename(ABLG = sample_id)

output %>%
  write_csv("csv_outputs/20231121_rockfishSPP_genetic_ids.csv")

output %>%
  ggplot(aes(x = z_score)) +
  geom_histogram() +
  facet_wrap(~repunit) +
  theme_bw()
```

Samples to add to the baseline:

```{r}
results_w_spp %>%
  filter(!str_detect(sample_id, "ABLG")) %>%
  filter(!str_detect(sample_id, "blue")) %>%
  mutate(species = gsub("\\d", "", sample_id)) %>%
  filter(!species %in% c("NTC-", "deacon")) %>%
  group_by(species) %>%
  ggplot(aes(x = z_score, fill = species)) +
  geom_histogram() +
  facet_wrap(~species) +
  theme_bw()

ggsave("pdf_outputs/norwegian_rockfish_zscores.pdf", width = 6, height = 3
)
```

